pyannote/audio/core/task.py

```diff
@@ -554,137 +530,142 @@ def get_smallest_type(value: int, unsigned: Optional[bool]=False) -> str:
+            annotated_regions, dtype=region_dtype
+        )
+        annotated_regions.clear()
+
+        self.prepared_data["annotations-segments"] = np.array(
+            annotations, dtype=segment_dtype
+        )
         annotations.clear()
-        prepared_data["metadata_unique_values"] = metadata_unique_values
 
-        if self.has_validation:
-            validation_chunks = list()
-
-            # obtain indexes of files in the validation subset
-            validation_file_ids = np.where(
-                prepared_data["metadata"]["subset"] == Subsets.index("development")
-            )[0]
-
-            # iterate over files in the validation subset
-            for file_id in validation_file_ids:
-                # get annotated regions in file
-                annotated_regions = prepared_data["annotated_regions"][
-                    prepared_data["annotated_regions"]["file_id"] == file_id
-                ]
+        self.prepared_data["metadata-values"] = metadata_unique_values
+        self.prepared_data["metadata-labels"] = np.array(
+            unique_labels, dtype=np.string_
+        )
+        unique_labels.clear()
+
+        self.prepared_data.update(self.prepare_validation())
+        self.prepared_data.update(self.post_prepare_data())
 
-                # iterate over annotated regions
-                for annotated_region in annotated_regions:
-                    # number of chunks in annotated region
-                    num_chunks = round(annotated_region["duration"] // duration)
-
-                    # iterate over chunks
-                    for c in range(num_chunks):
-                        start_time = annotated_region["start"] + c * duration
-                        validation_chunks.append((file_id, start_time, duration))
-
-            dtype = [
-                ("file_id", get_smallest_type(max(v[0] for v in validation_chunks))),
-                ("start", "f"),
-                ("duration", "f")
-            ]
-            prepared_data["validation_chunks"] = np.array(validation_chunks, dtype=dtype)
-            validation_chunks.clear()
-
-        self.prepared_data = prepared_data
         self.has_setup_metadata = True
 
         # save preparated data on the disk
-        if self.cache_path is not None:
-            with open(self.cache_path, 'wb') as cache_file:
-                np.savez_compressed(cache_file, **prepared_data)
+        if self.cache:
+            with open(self.cache, "wb") as cache_file:
+                np.savez_compressed(cache_file, **self.prepared_data)
 
         self.has_prepared_data = True
 
+    def post_prepare_data(self):
+        return dict()
+
     def setup(self):
         """Setup data on each device"""
-        # if all data was assigned to the task, nothing to do
+
         if self.has_setup_metadata:
+            # if metadata was already generated/assigned to the task, we stop here.
+            # this happens when ...
```

Comment:
this happens when setup is called on the main process, where `prepare_data` has already been run

Reply:
can you provide a sample script where this condition is true?

Reply:
This piece of code has been removed in [9e7abd3](https://github.com/pyannote/pyannote-audio/pull/1569/commits/9e7abd336882297aac42bb47d0cb430a61ed9508)